pkg/reconciler/space/reconciler.go

```diff
@@ -0,0 +1,248 @@
+)
+
+// Reconciler reconciles a Space object with the K8s cluster.
+type Reconciler struct {
+	*reconciler.Base
+
+	// listers index properties about resources
+	spaceLister     kflisters.SpaceLister
+	namespaceLister v1listers.NamespaceLister
+	roleLister      rbacv1listers.RoleLister
+}
+
+// Check that our Reconciler implements controller.Reconciler
+var _ controller.Reconciler = (*Reconciler)(nil)
+
+// Reconcile is called by Kubernetes.
+func (r *Reconciler) Reconcile(ctx context.Context, key string) error {
+	logger := logging.FromContext(ctx)
+
+	_, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	original, err := r.spaceLister.Get(name)
+	switch {
+	case apierrs.IsNotFound(err):
+		logger.Errorf("space %q no longer exists\n", name)
+		return nil
+
+	case err != nil:
+		return err
+
+	case original.GetDeletionTimestamp() != nil:
+		return nil
+	}
+
+	// Don't modify the informers copy
+	toReconcile := original.DeepCopy()
+
+	// Reconcile this copy of the service and then write back any status
+	// updates regardless of whether the reconciliation errored out.
+	reconcileErr := r.ApplyChanges(toReconcile)
+	if equality.Semantic.DeepEqual(original.Status, toReconcile.Status) {
+		// If we didn't change anything then don't call updateStatus.
+		// This is important because the copy we loaded from the informer's
+		// cache may be stale and we don't want to overwrite a prior update
+		// to status with this stale state.
+
+	} else if _, uErr := r.updateStatus(toReconcile); uErr != nil {
+		logger.Warnw("Failed to update Space status", zap.Error(uErr))
+		return uErr
+	}
+
+	return reconcileErr
+}
+
+// ApplyChanges updates the linked resources in the cluster with the current
+// status of the space.
+func (r *Reconciler) ApplyChanges(space *v1alpha1.Space) error {
+	space.Status.InitializeConditions()
+	namespaceName := resources.NamespaceName(space)
+
+	// Sync Namespace
+	{
+		desired, err := resources.MakeNamespace(space)
+		if err != nil {
+			return err
+		}
+
+		actual, err := r.namespaceLister.Get(desired.Name)
+		if errors.IsNotFound(err) {
+			actual, err = r.KubeClientSet.CoreV1().Namespaces().Create(desired)
+			if err != nil {
+				return err
+			}
+		} else if err != nil {
+			return err
+		} else if !metav1.IsControlledBy(desired, space) {
+			space.Status.MarkNamespaceNotOwned(namespaceName)
+			return fmt.Errorf("space: %q does not own namespace: %q", space.Name, namespaceName)
+		} else if actual, err = r.reconcileNs(desired, actual); err != nil {
+			return err
+		}
+
+		space.Status.PropagateNamespaceStatus(actual)
+	}
+
+	// If the namespace isn't ready (either it's coming up or being deleted)
+	// then this reconciliation process can't continue until we get notified it is.
+	if cond := space.Status.GetCondition(v1alpha1.SpaceConditionNamespaceReady); cond != nil {
+		if !cond.IsTrue() {
+			return fmt.Errorf("can't continue reconciling until namespace %q is ready", namespaceName)
```

Comment:
Is the contract here that only errors retrigger reconciliation? I'm wondering if we can return nil, since this is an expected path considering the eventual consistency.